Lazy init of ThreadWaitInfo in Thread.Mono.cs to prevent crash on external attached threads.

lateralusX opened the pull request:
Commit https://github.com/dotnet/runtime/commit/457f58cd30082c9de0be9ca62e64fa179c37dab3 added a new managed object into Mono's managed thread object, _waitInfo.

This new object was initialized in Thread managed constructor, problem with that is for all threads that gets their thread object created in native code, create_thread_object, that code did the same thing as constructor before commit, since managed 
constructor called down into native code calling the same native method as create_thread_object meaning a thread object created from managed or native would be initialized identical.

After above commit this is no longer true, meaning that all managed thread objects created in native code (like attached external threads) won't get _waitInfo initialized and that leads to a crash when OnThreadExiting gets called from Finalizer thread.

Fix is to make sure _waitInfo get initialize regardless of thread object gets created from managed or native code.

This fix implements a lazy init as part of property access, I tested a native version of the fix, calling back into managed to init managed parts of thread when thread gets created in native, but turns out that such a fix gets a little complicated, since some of the threads are created early when it's not possible to run managed code, so that in turn needed additional changes into the init order, that in turn has bigger impact and could cause other unwanted side effects.

Fixing it in managed using lazy/on demand init is simpler/cleaner approach but might come with a small performance hit for callers of WaitInfo property (will add volatile read and conditional branch), but looking at callers of the property, they appear in functions with rather high "cost", like Sleep, NewMutex, Wait, SignalAndWait, Interrupt and RelinquishOwnership.

//CC @CoffeeFlux

lateralusX committed:
Init ThreadWaitInfo to prevent crash on external attached threads.

Commit https://github.com/dotnet/runtime/commit/457f58cd30082c9de0be9ca62e64fa179c37dab3
added a new managed object into Mono's managed thread object, _waitInfo.

This new object was initialized in Thread managed constructor, problem
with that is for all threads that gets their thread object created in
native code, create_thread_object, that code did the same thing as
constructor before commit, since managed constructur called down into
native code calling the same native method as create_thread_object meaning
a thread object created from managed or native would be initialized
indentical.

After above commit this is no longer true, meaning that all managed thread
ojbects created in native code (like attached external threads) won't
get _waitInfo initialized and that leads to a crash when OnThreadExiting
gets called from Finalizer thread.

Fix is to make sure _waitInfo get initialize regardless of thread object
gets created from managed or native code.

This fix implements a lazy init as part of property access, I tested a
native version of the fix, calling back into managed to init managed
parts of thread when tread gets created in native, but turns out that
such a fix gets a little complicated, since some of the threads are
created early when it's not possible to run managed code, so that in turn
needed additional changes into the init order, that in turn has bigger
impact and could cause other unwanted side effects.

Fixing it in managed using lazy/on demand init is simpler/cleaner approach
but might come with a small performance hit for callers of WaitInfo property
(will add volatile read and conditional branch), but looking at callers
of the property, they appear in functions with rather high "cost",
like Sleep, NewMutex, Wait, SignalAndWait, Interrupt and
RelinquishOwnership.

vargaz commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -206,15 +205,15 @@ private void Initialize()
-            if (_waitInfo == null)
-                _waitInfo = new WaitSubsystem.ThreadWaitInfo(this);
+            return LazyInitializer.EnsureInitialized(ref _waitInfo, () => new WaitSubsystem.ThreadWaitInfo(this));
         }
```
This _might_ have perf problems, some hand written double locked code might be better.

lateralusX replied:
I looked at the implementation and fast path of EnsureInitialized and it does a volatile read and a branch:

```
public static T EnsureInitialized<T>([NotNull] ref T? target, Func<T> valueFactory) where T : class =>
            Volatile.Read(ref target!) ?? EnsureInitializedCore(ref target, valueFactory);
```

Only way to make faster is to do regular read, but then we are playing on the boundary of correctness since it will depend on compiler/hardware optimization, but since its a data dependency, I believe it will still be doing the right thing, at least on the CPU arch currently supported by runtime, but still, using volatile read will be more correct and using EnsureInitialized should be inlined into caller, so I believe this should be optimized into doing the same handwritten code, or did you have something else in mind?

lateralusX replied:
Also note that I only found references to the .WaitInfo property in functions that are quite expensive to begin with, like Sleep, NewMutex, Wait, SignalAndWait, Interrupt and RelinquishOwnership (called from SignalMutex).

jkotas replied:
The expensive part is the lamba allocation. The lambda `() => new WaitSubsystem.ThreadWaitInfo(this)` is going to run every time that `EnsureWaitInfo` is called. It is useless work 99+% of the time.

> using EnsureInitialized should be inlined into caller

I doubt it that EnsureInitialized be inlined with this change. The code to allocate lamda closure is pretty big.

lateralusX replied:
OK, well if that prevents inlining then I guess we need to manual inline the expression.

lateralusX replied:
Did a quick check on x64 (using CoreClr) after change, so the manual inlined code will be inlined into caller, looking like this:

lea         rax,[rcx+8]  
mov       rax,qword ptr [rax]  
test        rax,rax  
jne         00007FFAD41908C1  
call        00007FFAD41904D0  
mov      eax,3B9ACA00h

Since this is x64, the load volatile will just be a regular lea/move and then a conditional branch, all inlined into caller.

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -206,15 +205,15 @@ private void Initialize()
             InitInternal(this);
 #if TARGET_UNIX || TARGET_BROWSER
-            _waitInfo = new WaitSubsystem.ThreadWaitInfo(this);
+            EnsureWaitInfo();
```
Can this be delete now that the waitInfo is always allocate lazily?

lateralusX replied:
Jupp, we could eliminate that now.

lateralusX committed:
Review feedback, manual inline of EnsureInitialized. Running a similar
construct under CoreClr on x64 gives the following assembler on property
access:

lea   rax,[rcx+8]
mov   rax,qword ptr [rax]
test  rax,rax
jne   00007FFAD41908C1
call  00007FFAD41904D0
mov   eax,3B9ACA00h

Since its x64 fast path will be a lea/mov and conditional branch on
property access, all inlined in caller.

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -138,9 +138,8 @@ internal static int OptimalMaxSpinWaitsPerSpinIteration
-
 #if TARGET_UNIX || TARGET_BROWSER
-        internal WaitSubsystem.ThreadWaitInfo WaitInfo => _waitInfo;
+        internal WaitSubsystem.ThreadWaitInfo WaitInfo => EnsureWaitInfo();
```
```suggestion
        internal WaitSubsystem.ThreadWaitInfo WaitInfo => return Volatile.Read(ref _waitInfo) ?? AllocateWaitInfo();
```

lateralusX replied:
It was called from another place in code, so I kept it since it was inlined, but sure we can drop it and access property instread in InitializeCurrentThread.

jkotas replied:
Does that another place still need to call `EnsureWaitInfo` now that it is always allocated lazily?

lateralusX replied:
No, I dropped it in last commit.

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -199,22 +198,25 @@ public bool Join(int millisecondsTimeout)
 
 #if TARGET_UNIX || TARGET_BROWSER
-        private void EnsureWaitInfo()
+        private WaitSubsystem.ThreadWaitInfo EnsureWaitInfo()
+        {
+            return Volatile.Read(ref _waitInfo) ?? AllocateWaitInfo();
+        }
+
```
```suggestion
```

lateralusX replied:
Same comment as above.

lateralusX committed:
Drop EnsureWaitInfo and its usage.

jkotas commented on src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:
```diff
@@ -138,9 +138,8 @@ internal static int OptimalMaxSpinWaitsPerSpinIteration
-
 #if TARGET_UNIX || TARGET_BROWSER
-        internal WaitSubsystem.ThreadWaitInfo WaitInfo => _waitInfo;
+        internal WaitSubsystem.ThreadWaitInfo WaitInfo => Volatile.Read(ref _waitInfo) ?? AllocateWaitInfo();
```
```suggestion
        internal WaitSubsystem.ThreadWaitInfo WaitInfo
        {
            return Volatile.Read(ref _waitInfo) ?? AllocateWaitInfo();
            
            WaitSubsystem.ThreadWaitInfo AllocateWaitInfo()
            {
                Interlocked.CompareExchange(ref _waitInfo, new WaitSubsystem.ThreadWaitInfo(this), null);
                return _waitInfo;
            }
        }
```
Nit: To keep the related code together and reduce number of ifdefs.

jkotas replied:
(Also delete the other AllocateWaitInfo. github does not let me to do it as suggestion.)

jkotas reviewed (approved):
LGTM otherwise

CoffeeFlux reviewed (approved):
Generally LGTM—appreciate you looking into this!

lateralusX committed:
Reduce number of ifdefs.
